source/common/stats/stats_impl.cc

```diff
@@ -256,18 +267,26 @@ TagProducerImpl::addDefaultExtractors(const envoy::config::metrics::v2::StatsCon
 }
 
 void HeapRawStatDataAllocator::free(RawStatData& data) {
-  // This allocator does not ever have concurrent access to the raw data.
-  ASSERT(data.ref_count_ == 1);
+  ASSERT(data.ref_count_ > 0);
+  if (--data.ref_count_ > 0) {
+    return;
+  }
+
+  std::unique_lock<std::mutex> lock(mutex_);
```

Comment:
The problem isn't the stat, it's the set.

Reply:
+1.  Also, side note: that comment is no longer valid since the same stat can be freed/allocated multiple times, meaning that there may be cases where the allocator is operating on the same raw stat from two different threads.

Reply:
AFAICT we only ever do these allocations under existing locking, e.g. https://github.com/envoyproxy/envoy/blob/872728d4ad50acececaeae389da59cb72846cf89/source/common/stats/thread_local_store.cc#L192. Do we need to be double locking here?

I might be wrong in my assessment, please point out if not (and add a comment to the code!).

Reply:
Yes, there probably need to be comments around this.  The `alloc()` calls are protected, but the `free()` calls are made from the destructors of the individual stat objects.  See https://github.com/envoyproxy/envoy/blob/master/source/common/stats/stats_impl.h#L310 for an example.